runtime/stdlib/test.go

```diff
@@ -288,6 +293,61 @@ func init() {
 		),
 	)
 
+	// Test.beEmpty()
```

Comment:
Could these new functions also just be declared in Cadence (`test.cdc`)? Ideally we would implement as much as possible in Cadence itself.

Not sure why `equal` was already implemented in Go, it looks implementable in Cadence, too.

Reply:
Not all types are equatable at the Cadence level. So `equal` method had to be implemented in Go.

But yeah, +1 for implementing these matchers as much as possible in Cadence, if possible.

Reply:
I made a first attempt to port all of the 5 remaining matchers in Cadence:

```cadence
pub fun beLessThan(_ value: Number): Matcher {
    return Matcher(test: fun (testValue: AnyStruct): Bool {
        if let otherValue = testValue as? Number {
            return otherValue < value
        }
        return false
    })
}
```
**fails with:**
```bash
panic: Checking failed:
error: cannot apply binary operation < to types: `Number`, `Number`
```
---

```cadence
pub fun beGreaterThan(_ value: Number): Matcher {
    return Matcher(test: fun (testValue: AnyStruct): Bool {
        if let otherValue = testValue as? Number {
            return otherValue > value
        }
        return false
    })
}
```
**fails with:**
```bash
panic: Checking failed:
error: cannot apply binary operation > to types: `Number`, `Number`
```
---

```cadence
pub fun beEmpty(): Matcher {
    return Matcher(test: fun (testValue: AnyStruct): Bool {
        if let arr = testValue as? [AnyStruct] {
            return arr.length == 0
        } else if let dict = testValue as? {AnyStruct: AnyStruct} {
            return dict.length == 0
        }
        return false
    })
}
```
**fails with:**
```bash
panic: Checking failed:
error: cannot use type as dictionary key type: `AnyStruct`
```
---

```cadence
pub fun haveElementCount(_ count: Int): Matcher {
    return Matcher(test: fun (testValue: AnyStruct): Bool {
        if let arr = testValue as? [AnyStruct] {
            return arr.length == count
        } else if let dict = testValue as? {AnyStruct: AnyStruct} {
            return dict.length == count
        }
        return false
    })
}
```
**fails with:**
```bash
panic: Checking failed:
error: cannot use type as dictionary key type: `AnyStruct`
```
---

```cadence
pub fun contain(_ element: AnyStruct): Matcher {
    return Matcher(test: fun (testValue: AnyStruct): Bool {
        if let arr = testValue as? [AnyStruct] {
            return arr.contains(element)
        } else if let dict = testValue as? {AnyStruct: AnyStruct} {
            return dict.containsKey(element)
        }
        return false
    })
}
```
**fails with:**
```bash
panic: Checking failed:
error: cannot compare value which has type: `AnyStruct`

error: cannot use type as dictionary key type: `AnyStruct`
```

Any workarounds for these errors?

Reply:
Thank you for giving it a shot! 👏 

- Supporting comparisons of `Number`s, i.e. potentially different kinds of numbers, is probably not something we would want to add. You could extend the implementation of the matcher to try and downcast to each concrete number type (e.g. cast LHS and RHS to `UInt8` and compare, then to `UInt16`, and so on), but there are a lot of number types.
- Currently there is no type exposed to Cadence programs which represents the type that can be used as a `Dictionary` key type. Basically, we would want to introduce a "`HashableStruct`" interface, and have all types mentioned in `IsValidDictionaryKeyType` conform to it.

In general, given the Cadence implementation are not (yet) possible, let's just keep the Go implementations for now 👍